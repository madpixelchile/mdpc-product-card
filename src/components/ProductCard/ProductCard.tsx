
import React,{ createContext } from 'react';
import styles from '../../styles/styles.module.css';

import { useProduct } from '../../hooks/useProduct';
import { Product, ProductContextProps, OnChangeArgs, ProductCartInitialValues, ProductCardHandlers } from '../../interfaces/ProductInterfaces';

export const ProductContext = createContext({} as ProductContextProps);
const { Provider } = ProductContext;

export interface Props {
    product: Product;
    children: ( args: ProductCardHandlers ) => React.JSX.Element,
    className?: string;
    style?: React.CSSProperties;
    onChangeProduct?: (args:OnChangeArgs)=> void;
    value?: number;
    initialValues?: ProductCartInitialValues;
}

export const ProductCard = ({ children, product, className = '', style, onChangeProduct, value, initialValues }: Props) => {

    const { counter, increaseBy, maxCount, isMaxCountReached, reset } = useProduct({onChangeProduct, product, value, initialValues });

    return (
        <Provider
            value={{
                counter,
                increaseBy,
                product,
                className,
                maxCount
            }}
        >
            <div 
                className={`${styles.productCard} ${className}`} 
                style={style}
            >
                {
                    children({
                        count: counter,
                        isMaxCountReached,
                        product,
                        maxCount,
                        increaseBy,
                        reset,
                    }) 
                }
            </div>
        </Provider>
    )
}
