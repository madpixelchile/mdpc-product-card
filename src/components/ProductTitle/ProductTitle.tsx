import React, { useContext } from "react";
import { ProductContext } from "../ProductCard/ProductCard";
import styles from '../../styles/styles.module.css';

export interface Props{
    title?: string;
    className?: string;
    style?: React.CSSProperties;
}

export const ProductTitle = ({ title,  className = '', style }: Props ) => {
    const { product } = useContext(ProductContext);
    let productTitle: string;
    if(title){
        productTitle = title;
    }else if(product.title){
        productTitle = product.title;
    }else{
        productTitle = ''
    }
    return (
        <span className={`${styles.productDescription} ${className}`} style={style}>{ productTitle }</span>
    )
}



