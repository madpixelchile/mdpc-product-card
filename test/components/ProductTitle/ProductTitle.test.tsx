
import React from "react";

import renderer from 'react-test-renderer'
import { ProductTitle } from '../../../src/components/ProductTitle/ProductTitle';
import { ProductCard } from "../../../src/components/ProductCard/ProductCard";

import { product1, product2 } from "../../data/products";
import { ProductImage } from "../../../src/components/ProductImage/ProductImage";


describe('Product Title', () => {

    test('Debe de mostrar el componente correctamente con el título personalizado.', () => {

        const wrapper = renderer.create(
            <ProductTitle title={'Custom Product'} style={{ color: 'blue' }} />
        )

        expect(wrapper.toJSON()).toMatchSnapshot();

    })

    test('El producto no debe de mostrar la imagen cuando no está en el objeto', () => {

        const productCard = renderer.create(
            <ProductCard product={product1}>
                {
                    () => (
                        <>
                            <ProductTitle />
                        </>
                    )
                }
            </ProductCard>
        )

        expect(productCard.toJSON()).toMatchSnapshot();

    })

    test('El producto debe de mostrar una imagen', () => {

        const productCard = renderer.create(
            <ProductCard product={product2}>
                {
                    () => (
                        <>
                            <ProductTitle/>
                            <ProductImage/>
                        </>
                    )
                }
            </ProductCard>
        )

        expect(productCard.toJSON()).toMatchSnapshot();

    })

});
