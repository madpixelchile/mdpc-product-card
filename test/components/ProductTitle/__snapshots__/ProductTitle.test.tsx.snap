// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Title Debe de mostrar el componente correctamente con el título personalizado. 1`] = `
<span
  className="productDescription "
  style={
    Object {
      "color": "blue",
    }
  }
>
  Custom Product
</span>
`;

exports[`Product Title El producto debe de mostrar una imagen 1`] = `
<div
  className="productCard "
>
  <span
    className="productDescription "
  >
    Producto dos
  </span>
  <img
    alt="./coffee-mug2.png"
    className="productImg "
    src="./coffee-mug2.png"
  />
</div>
`;

exports[`Product Title El producto no debe de mostrar la imagen cuando no está en el objeto 1`] = `
<div
  className="productCard "
>
  <span
    className="productDescription "
  >
    Coffee Mug - Sin Imagen
  </span>
</div>
`;
